File: 112-path-sum/path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public boolean DFS(TreeNode node, int targetSum, int findSum){
        //if node doesn't exist return false
        if(node == null){
            return false;
        }   

        // update current sum
        findSum += node.val;

        // check if we are at a leaf node
        if (node.left == null && node.right == null){
            return targetSum == findSum; // check if equals
        }

        // else keep exploring left or right tree
        return DFS(node.left,targetSum,findSum) || DFS(node.right,targetSum,findSum);
    }

    public boolean hasPathSum(TreeNode root, int targetSum) {
        int findSum = 0;
        return DFS(root,targetSum,findSum);
    }

    // T O(n) S O(h)
}

File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        """
        self.dfs(root.left,result)
        """
        result = 0
        return self.dfs(root, result, targetSum)

    def dfs(self, root, result, targetSum): # inorder traversal, go as far left, then check right, backtrack
        if root is None:
            return False
        # [5, 4, 11, 2]
        result += root.val

        if not root.right and not root.left: # if a leaf node
            print(f"{root.val} is a leaf, current result is {result} target is {targetSum}")
            return result == targetSum

        return self.dfs(root.left, result, targetSum) or self.dfs(root.right, result, targetSum)
        
        

            
        